### PR TITLE
feat(symbolsources): Add explicit API to configure symbol sources

### PR DESCRIPTION
**Pull request by flub**
This is the base of adding an explicit API to handle symbol sources
instead of having to use the project details for this.  It moves more
of the logic to the backend and provides a more natural API which does
not require manually manipulating JSON-encoded strings.

### CODE REVIEW

File: src/sentry/api/endpoints/project_symbolsources.py [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+"""
+
+from rest_framework.request import Request
+from rest_framework.response import Response
+
+from sentry.api.bases.project import ProjectEndpoint, StrictProjectPermission
+from sentry.models import Project
+
+
+class ProjectSymbolSourcesEndpoint(ProjectEndpoint):  # type: ignore
```
Comment (wedamija) [inline src/sentry/api/endpoints/project_symbolsources.py:27]: Nit: A common pattern throughout our api is that we call the endpoint that operates on a set of data the `index` endpoint and the specific endpoint the `detail` endpiont, and keep these in separate files.

So we'd call this `ProjectSymbolSourcesIndexEndpoint` and put in `src/sentry/api/endpoints/project_symbol_sources_index.py`
and the other `ProjectSymbolSourcesDetailEndpoint` and put in `src/sentry/api/endpoints/project_symbol_sources_details.py`

Comment (flub) [reply src/sentry/api/endpoints/project_symbolsources.py]: ❤️  Thanks for the pointer!  I've updated this.

**Review by wedamija — APPROVED** [review]
LGTM, thanks for working on this, the more we can pull out of the project serializer the better.

**Comment by flub** [comment]
@wedamija one question that came up in relation to this was whether we're actually allowed to remove the `symbolSources` property from the project details endpoint once this is implemented or whether project details is supposed to be a faithful representation of the project options.  I would consider that the storage of this data inside project options is no more than an internal implementation detail at that point and no longer expose it on the project details, but is that bad practice or encouraged?

**Comment by wedamija** [comment]
> @wedamija one question that came up in relation to this was whether we're actually allowed to remove the `symbolSources` property from the project details endpoint once this is implemented or whether project details is supposed to be a faithful representation of the project options. I would consider that the storage of this data inside project options is no more than an internal implementation detail at that point and no longer expose it on the project details, but is that bad practice or encouraged?

I think that's fine. We don't have a specific practice here, but there's no expectation that an endpoint is a perfect representation of the underlying data (like you said, that's just an implementation detail). One thing this does raise is that maybe `symbolSources` should be located in their own table, but I assume that's a lot of work for little gain at this point.

The main thing to consider here is whether we'd expect external users of our api to be using these values. If this is really obscure and only used by sentry then there's no problem. If others might be using it then we will need to think through how to properly deprecate this value from that api.

**Comment by flub** [comment]
>  One thing this does raise is that maybe `symbolSources` should be located in their own table, but I assume that's a lot of work for little gain at this point.

Yes, this could indeed be argued.  But even worse, because of the possible concurrency that this API introduces on this array stored inside a project option maybe it even is a requirement.  Though if you simply use our UI this won't be a problem because the way to change these things is modal and you wouldn't create concurrent updates.

> The main thing to consider here is whether we'd expect external users of our api to be using these values. If this is really obscure and only used by sentry then there's no problem. If others might be using it then we will need to think through how to properly deprecate this value from that api.

My judgement is that is is pretty obscure and never been documented so I expect we'll be fine with just removing it once the UI no longer uses it.

**Comment by wedamija** [comment]
I just removed my approval since this is still in draft and being worked on. If you want feedback on your changes in the draft let me know, otherwise happy to review once it's ready to go.

**Comment by flub** [comment]
Future memo: this api should also handle cases where the user no longer has the custom symbol sources feature flag and has several configured-but-disabled repositories.  See e.g. https://github.com/getsentry/sentry/pull/31107#issuecomment-1012973534

**Pull request closed without merging**